service/history/workflow/mutable_state_impl.go

```diff
@@ -4236,6 +4236,82 @@ func (ms *MutableStateImpl) RejectWorkflowExecutionUpdate(_ string, _ *updatepb.
+	}
+	ms.GetExecutionInfo().VersioningInfo.VersioningOverride = override
+
+	// If effective deployment or behavior change, we need to reschedule any pending tasks, because History will reject
+	// the task's start request if the task is being started by a poller that is not from the workflow's effective
+	// deployment according to History. Therefore, it is important for matching to match tasks with the correct pollers.
+	// Even if the effective deployment does not change, we still need to reschedule tasks into the appropriate
+	// default/unpinned queue or the pinned queue, because the two queues will be handled differently if the task queue's
+	// Current Deployment changes between now and when the task is started.
+	//
+	// We choose to let any started WFT that is running on the old deployment finish running, instead of forcing it to fail.
+	if !proto.Equal(ms.GetEffectiveDeployment(), previousCurrentDeployment) ||
+		ms.GetEffectiveVersioningBehavior() != previousEffectiveVersioningBehavior {
+		// If there is an ongoing transition, we remove it so that tasks from this workflow (including the pending WFT
+		// that initiated the transition) can run on our override deployment as soon as possible.
+		//
+		// We only have to think about the case where the workflow is unpinned, since if the workflow is pinned, no
+		// transition will start.
+		//
+		// If we did NOT remove the transition, we would have to keep the pending WFT scheduled per the transition's
+		// reschedule, so that when the task is started it can run on the transition's target deployment, complete,
+		// and thereby complete the transition. If there is anything wrong with the transition's target deployment,
+		// the transition could hang due to the task being stuck, or the transition could fail if the WFT fails.
+		// Instead of waiting for that, we remove the transition and all is well.
+		//
+		// Once the override is set, any attempts to start a transition will be rejected.
+		//
+		// It is possible for there to be an ongoing transition and an override that both result in the same effective
+		// behavior and effective deployment. In that case, we would not hit the code path to remove the transition or
+		// reschedule the WFT. For this to happen, the existing behavior and the override would both have to be unpinned.
+		// If we don't remove the transition or reschedule pending tasks, the outstanding WFT on the transition's
+		// target queue will be started on the transition's target deployment, it will likely succeed, since the override
+		// also points to that deployment, so the transition will complete successfully.
+		// If we removed the transition, the pending WFT would try to start on the transition's target deployment, but
+		// the effective deployment in history would not match, so the task would be rejected and rescheduled.
+		// So we should keep the transition to avoid an extra reschedule cycle.
```

Comment:
Let's change it so it doesn't look like we're assuming the existing transition will complete.

```suggestion
		// target queue will be started on the transition's target deployment. Most likely this matches user's intention because they add the unpinned override, they want to workflow to do the transition. Even if we removed the transition, the rescheduled task will be redirected by Matching to the old transition's deployment again and it will start the same transition in the workflow. So removing the transition does not make a difference and just adds some extra work for the server.
```